### PR TITLE
pod-utils: upload artifacts dir and fix bug in gcsupload

### PR DESCRIPTION
**Pull request by cjwagner**
The pod utilities are working for us now!
cc @steuhs 
/assign @stevekuznetsov 
/area prow
/kind bug

### CODE REVIEW

File: prow/entrypoint/options.go [hunk 1]
```diff
@@ -45,6 +45,11 @@ type Options struct {
 	// sending SIGINT before the entrypoint sends
 	// SIGKILL.
 	GracePeriod time.Duration `json:"grace_period"`
+	// ArtifactDir is a directory where test processes can dump artifacts
+	// for upload to persistent storage (courtesy of sidecar).
+	// If specified, it is created by entrypoint before starting the test process.
+	// May be ignored if not using sidecar.
+	ArtifactDir string `json:"artifact_dir,omitempty"`
```
Comment (stevekuznetsov) [inline prow/entrypoint/options.go]: The design before was that we would pass `$ARTIFACTS` to the test binary to signify this. Should be fine to also have the entry point take it explicitly, but then we should not set it on the container and have the entry point set the env explicitly for the wrapped test process so we have just the one source of truth. We should also doc this so test authors know what to expect

File: prow/entrypoint/run.go [hunk 2]
```diff
@@ -102,6 +116,9 @@ func (o Options) Run() error {
 		cancelled = true
 		gracefullyTerminate(command, done, gracePeriod)
 	}
+	// Close the process logfile before writing the marker file to avoid racing
+	// with the sidecar container.
+	processLogFile.Close()
```
Comment (stevekuznetsov) [inline prow/entrypoint/run.go]: Nice

File: prow/errorutil/aggregate.go [hunk 3]
```diff
@@ -18,21 +18,23 @@ package errorutil
 
 import (
 	"fmt"
+	"strings"
```
Comment (stevekuznetsov) [inline prow/errorutil/aggregate.go]: This used to be a copy paste of utils in `k/k` I couldn't figure out how to import... I don't think they're in any staging repo. Anyway, I don't think it matters if we fork this but just FYI

File: prow/entrypoint/options.go (hunk 1 above)
Comment (cjwagner) [reply prow/entrypoint/options.go]: I don't think that Steve is suggesting we change the interface exposed to jobs. If I understand correctly, jobs will still receive the path to the artifacts dir via the `$ARTIFACTS` envvar, but the plumbing will be a bit different.
As this PR is now, the podspec decoration specifies the artifact path in the `$ARTIFACTS` envvar and as a field in the entrypoint configuration envvar creating two sources of truth (albeit consistent ones). I think Steve is suggesting that we only specify the path in the entrypoint config and then have the entrypoint binary augment the env with `$ARTIFACTS` based on its config before execing the test binary.

Comment (stevekuznetsov) [reply prow/entrypoint/options.go]: Yes, that is what I meant

Comment (cjwagner) [reply prow/entrypoint/options.go]: I was just working on cleaning this up, but I thought of one reason we might not want to have the entrypoint binary set the ARTIFACTS env var at runtime:
When specifying args to the test container in the ProwJob config, environment variables can be specified with `$(ENV_VAR)` syntax and will be expanded before the container runs. However, if we have the entrypoint binary set the ARTIFACTS env var after the container starts it won't be possible to expand the ARTIFACTS env var in the test container args. Test binaries will either have to consume the env var directly or be wrapped with a shell call that expands the environment variables (perhaps the pod utilities should do this automatically?).

**Pull request merged**